reports/helpers.py

```diff
@@ -0,0 +1,261 @@
+        - Today if this is the current month and the server is not terminated yet.
+        - The end of the billing month if the server terminated after that month.
+        - The termination date of the server if it's terminated before the end of this month.
+
+    :param appserver: An OpenEdxAppServer instance.
+    :param invoice_date: The beginning of the month we want to fetch dates for.
+    :return: A pair of first and last billing days of the month
+    """
+    today = timezone.now()
+    invoice_month = _normalize_date(invoice_date)
+
+    future_billing_month = invoice_month > today
+    appserver_terminated_before = appserver.terminated and appserver.terminated < invoice_month
+    appserver_created_after = _normalize_date(appserver.created) > invoice_month
+
+    if future_billing_month or appserver_terminated_before or appserver_created_after:
+        return None, None
+
+    if _normalize_date(today) == invoice_month:
+        end_of_month = today
+    else:
+        _, month_days = calendar.monthrange(invoice_month.year, invoice_month.month)
+        end_of_month = invoice_month + timedelta(days=month_days - 1)
+
+    terminated_after_billing_month = appserver.terminated and _normalize_date(appserver.terminated) > invoice_month
+    if not appserver.terminated or terminated_after_billing_month:
+        last_billing_day = end_of_month
+    else:
+        last_billing_day = appserver.terminated
+
+    if appserver.created < invoice_month:
+        first_billing_day = invoice_month
+    else:
+        first_billing_day = appserver.created
+
+    return first_billing_day, last_billing_day
+
+
+def generate_charge_details(appserver, invoice_period):
+    """
+    Generates all of the charges and their details in a given period for a
+    specific AppServer
+    :param appserver: An OpenEdxAppServer
+    :param invoice_period: The beginning of the month we want to issue usages for.
+    :return: A dictionary all charges details. Example:
+                {
+                    'name': 'AppServer 1',
+                    'billing_start': 1 - 9 - 2018,
+                    'billing_end': 30 - 9 - 2018,
+                    'days': 30,
+                    'charge': 300,
+                }
+    """
+    days = 0
+    charge = 0
+    billing_start, billing_end = get_billing_dates(appserver, invoice_period)
+
+    if billing_start and billing_end:
+        # Adding one to the days to include the first day
+        days = billing_end.day - billing_start.day + 1
+        charge = settings.BILLING_RATE * days
+
+    return {
+        'name': appserver.name,
+        'billing_start': billing_start,
+        'billing_end': billing_end,
+        'days': days,
+        'charge': charge,
+    }
+
+
+def get_instance_charges(instance, invoice_period):
+    """
+    Generates the charges for the given instance's terminated and running
+    AppServers in the given period of the invoice date.
```

Comment:
„in the given period of the invoice date“: That's not clear. `invoice_period` isn't a period (it's not a tuple which specifies a start and an end), it's a date but it represents a month. All 3 names are confusing.
You could just leave it as „in the given month“.

Reply:
True, I only tried to get a good name for it to not make it confusing more as it's a date in a specific month of an invoice that's going to be normalized. But I was thinking since the view is going to generate a `datetime` object for the invoice start date, we can name it `invoice_start_date` which might reduce the confusion.